### PR TITLE
Harden unauthenticated post rating handlers against IDOR and abuse

### PR DESCRIPTION
**Pull request by ShubhamGupta05**
## Summary

Hardens the two **unauthenticated** rating AJAX handlers (`bsf_submit_rating` / `bsf_update_rating` in `functions.php`). Before this change they wrote `post-rating` post-meta using a client-supplied `post_id` with no validation, trusted a client-supplied IP, did no server-side dedup, and did not bound the rating value — allowing rating forgery on arbitrary posts and unbounded post-meta writes.

- **Object validation** — new `bsf_is_rateable_post()` gate: only **published** posts whose schema type actually renders a rating form (Product `6`, Recipe `7`, Software `8`) are accepted. Closes the arbitrary / draft / non-existent post-meta write (IDOR).
- **Server-side IP** — the visitor IP is now derived via `get_the_ip()` instead of the posted `ip` field, so the dedup key can't be forged from the request body.
- **One rating per IP per post** — enforced server-side to block duplicate and unbounded submissions.
- **Range check** — rating values outside `1–5` are rejected, preventing aggregate-rating skew.

Scoped, proportionate change — no rework of the rating storage model or IP resolution (see note below).

## Known residual (intentional, out of scope here)

`get_the_ip()` still honours `X-Forwarded-For` / `Client-IP` for proxy/CDN compatibility, so a determined attacker rotating spoofed headers can still register more than one vote on a genuinely-published rateable post. Fully closing this means resolving IP from `REMOTE_ADDR`, which breaks dedup accuracy behind reverse proxies/CDNs — deliberately left for a separate discussion.

## Test plan

Verified end-to-end on a local install (unauthenticated, against `admin-ajax.php`) and with an isolated unit harness on the extracted handlers:

- [x] Legit rating on a published Product stores exactly one row.
- [x] A spoofed `ip` in the POST body is ignored; the server-derived IP is stored.
- [x] A second submission from the same visitor is rejected ("already rated") — no extra row.
- [x] Submissions targeting a draft / non-rateable / non-existent post ID are rejected — no write.
- [x] Out-of-range and zero star values are rejected.
- [x] Invalid nonce is rejected.
- [x] Update handler updates the caller's own row in place; a new IP adds its own row.
- [x] `php -l` clean on `functions.php`.

## Notes

- Adds a `1.7.9` changelog entry in `readme.txt` and `README.md`.
- New helper documented with `@since 1.7.9`.

### CODE REVIEW

**Comment by ShubhamGupta05** [comment]
## Tester QA — rating handler hardening

Fix affects the unauthenticated rating AJAX endpoints `bsf_submit_rating` / `bsf_update_rating`. Below are functional + security test cases. **On this PR branch you should see the "Expected (FIXED)" results.** To confirm the "before" behaviour, run the same steps on `dev` and restart the site's PHP (OPcache) between branch switches.

### Setup (once)
1. Create/pick a **published** post → in the "Configure Rich Snippet" meta box set **Type = Product** (Recipe/Software also work), add a name + price, Update.
2. Open it on the front end — the **star rating widget** must be visible. Note its **post ID**.
3. Also note any **plain post/page ID** (no rating) to use as an IDOR target.

> The widget only renders for Product/Recipe/Software — that is by design and is the anchor for the fix.

---

### A. Functional (UI — logged out)
| # | Steps | Expected (FIXED) |
|---|---|---|
| A1 | Open the Product post, click a star (e.g. 4) | "Ratings added successfully !"; reload; aggregate reflects the vote |
| A2 | Same visitor rates again | "You have already rated this item." — count does **not** increase |
| A3 | From a different device/network (different IP), rate the same post | Accepted; count +1 (legit multi-user still works) |
| A4 | Repeat on a Recipe and a Software post | Works identically |

---

### B. Security / abuse (unauthenticated — Terminal / curl)

**B0 — scrape the public nonce (attacker's first step):**
```bash
SITE="http://localhost:10069"          # your site URL
RATEABLE_ID=1                          # published Product/Recipe/Software post ID
NONCE=$(curl -sL "$SITE/?p=$RATEABLE_ID" | grep -oE 'name="bsf_rating_nonce" value="[a-f0-9]+"' | head -1 | grep -oE '[a-f0-9]+"$' | tr -d '"')
AJAX="$SITE/wp-admin/admin-ajax.php"
echo "nonce=$NONCE"
```
If `nonce=` is empty, the post is not published or not set to Product/Recipe/Software.

| # | Command (paste, then read the JSON) | Expected (FIXED) | Before-fix (VULNERABLE) |
|---|---|---|---|
| B1 Legit add | `curl -s "$AJAX" -d "action=bsf_submit_rating&bsf_rating_nonce=$NONCE&post_id=$RATEABLE_ID&star-review=5&ip=9.9.9.9"; echo` | `success:true`; stored IP is the **real server IP, not 9.9.9.9** | Accepted; stores attacker's `9.9.9.9` |
| B2 Duplicate | repeat B1 | `success:false` "You have already rated this item." | Accepted again — a vote every call |
| B3 IDOR non-existent | `curl -s "$AJAX" -d "action=bsf_submit_rating&bsf_rating_nonce=$NONCE&post_id=999999&star-review=5&ip="; echo` | `success:false` "Invalid rating request." — no row written | Writes rating meta to a fake ID |
| B4 IDOR non-rateable | same as B3 but `post_id=<plain post/page id>` | `success:false` "Invalid rating request." | Writes rating meta to an arbitrary post |
| B5 IDOR draft | set a Product post to **Draft**, use its ID | `success:false` "Invalid rating request." | Writes to an unpublished post |
| B6 Out-of-range | `...&post_id=$RATEABLE_ID&star-review=1000000&ip=` | `success:false` "Invalid rating value." | Accepted — skews the aggregate |
| B7 Zero stars | `...&star-review=0&ip=` | `success:false` "Invalid rating value." | Accepted as 0 |
| B8 Bad nonce | `...&bsf_rating_nonce=bad&post_id=$RATEABLE_ID&star-review=5` | `0` (rejected) | `0` (unchanged) |
| B9 Update own row | `curl -s "$AJAX" -d "action=bsf_update_rating&bsf_rating_nonce=$NONCE&post_id=$RATEABLE_ID&star-review=3&ip="; echo` (after B1, same IP) | `success:true` "updated", row updated in place (count unchanged) | Overwrites the first stored row regardless of owner |

**Verify what was written (DB):**
```sql
SELECT post_id, meta_value FROM wp_postmeta
WHERE meta_key='post-rating' AND post_id IN (RATEABLE_ID, 999999, PLAIN_POST_ID);
```
FIXED: nothing on 999999/plain post; only one row per real IP on the Product. VULNERABLE: rows on 999999 + many spoofed-IP rows.

**Cleanup:**
```sql
DELETE FROM wp_postmeta WHERE meta_key='post-rating' AND post_id IN (RATEABLE_ID, 999999, PLAIN_POST_ID);
```

---

### C. Known residual — expected, NOT a bug for this PR
`get_the_ip()` still honours `X-Forwarded-For` (proxy/CDN compatibility), so rotating that header can still register multiple votes **on a genuinely published rateable post**:
```bash
curl -s "$AJAX" -H "X-Forwarded-For: 5.5.5.5" -d "action=bsf_submit_rating&bsf_rating_nonce=$NONCE&post_id=$RATEABLE_ID&star-review=5&ip="; echo
```
Deliberate tradeoff, out of scope here. **Only report deviations from the "Expected (FIXED)" column.**

### Pass criteria
A1–A4 as described, and B1–B9 match **Expected (FIXED)** — most importantly: B3/B4/B5 rejected with **no DB write**, B1 stores the **server IP not 9.9.9.9**, B2 rejected as duplicate, B6/B7 rejected.

### Notes for reproducing "before"
- Switch to `dev`, then **Stop/Start the site in Local** so PHP OPcache reloads the old bytecode (otherwise the fixed code keeps serving and you'll wrongly see the FIXED results).

**Review by sushmak02 — APPROVED** [review]

**Comment by sushmak02** [comment]
## Review notes

The security hardening here is well-targeted and mostly correct, but the second commit's `get_the_ip()` rewrite breaks the rating feature outright on any CDN-fronted site. Flagging that as blocking. Heads-up that the PR already carries an approval at `1ac332f` — the issue below landed in that same commit, so it's worth a second look before merge.

### Supply-chain checklist

Clean. Both commits signed (`verified: true`), GitHub actor matches the author on both, no burst-push pattern. The diff touches only `functions.php`, `README.md`, `readme.txt` and the `.pot` — no build config. I also content-scanned the branch head's `Gruntfile.js`, `package.json` and `.gitignore` for the known payload signatures: no matches (longest `Gruntfile.js` line is 83 chars). No dangerous PHP sinks introduced. `php -l` clean on `functions.php`.

---

### Blocking — `get_the_ip()` collapses all visitors to one identity behind a CDN

`functions.php:1203`. The proxy heuristic is "`REMOTE_ADDR` is private/reserved ⇒ we're behind a proxy". A CDN edge connects from a **public** address, so it is never detected. Running the new function verbatim against real topologies (two distinct visitors, A=`203.0.113.10`, B=`198.51.100.77`):

```
1a direct, visitor A, no spoof          got=203.0.113.10  want=203.0.113.10  OK
1b direct, visitor A spoofs XFF         got=203.0.113.10  want=203.0.113.10  OK
2a local proxy (nginx to fpm)           got=203.0.113.10  want=203.0.113.10  OK
2b local proxy, visitor A spoofs XFF    got=203.0.113.10  want=203.0.113.10  OK
3a Cloudflare, visitor A                got=172.68.44.9   want=203.0.113.10  *** MISMATCH ***
3b Cloudflare, visitor B (same edge)    got=172.68.44.9   want=198.51.100.77 *** MISMATCH ***
4a CDN + private LB, visitor A          got=172.68.44.9   want=203.0.113.10  *** MISMATCH ***
4b CDN + private LB, visitor B          got=172.68.44.9   want=198.51.100.77 *** MISMATCH ***
```

Scenario 3: `$behind_proxy` is false because the CF edge IP is public, so the forwarded header is discarded and every visitor is keyed to the edge IP. Scenario 4: `$behind_proxy` is true, but rightmost-XFF yields the address the load balancer saw — again the edge IP.

Combined with the new one-rating-per-IP rule, the user-visible result on a CDN site is:

- The first visitor rates the post. Every later visitor hits `bsf_do_rating()`, finds "their" IP already in the array, and is served `bsf_display_rating()` with **a stranger's stars pre-selected**.
- Their submit routes to `bsf_update_rating()`, which overwrites that shared row — so the feature degrades to a single globally-mutable rating.
- `rating_count()` stays at 1 forever, so the emitted `aggregateRating` reports 1 review regardless of traffic.

This is a regression: before the change XFF was trusted unconditionally, so CDN sites resolved distinct visitors correctly (spoofable, but functional). The changelog line *"Sites behind a reverse proxy or CDN are detected automatically"* isn't accurate for the CDN case and shouldn't ship as written.

The escape hatch does partly work — on scenario 3, `add_filter( 'bsf_trust_forwarded_ip', '__return_true' )` resolves correctly, because Cloudflare appends the client IP to any incoming XFF, making rightmost genuinely unspoofable. It doesn't rescue scenario 4, which needs a trusted-hop count. Suggested direction: derive the client from an explicitly configured trusted header or trusted-proxy CIDR list with a hop count, rather than inferring proxy-ness from private ranges.

### Medium — private posts are still writable by unauthenticated callers

`bsf_is_rateable_post()` at `functions.php:1356` admits `private` alongside `publish`, but there's no `current_user_can( 'read_post', $post_id )` anywhere in the rating path — there are zero capability checks in the file. The nonce action is the bare string `bsf_rating`, not post-scoped, so a nonce lifted from any public rateable post replays against a private post ID. An anonymous caller who guesses a private post's ID can write rating meta to content they can't read.

The first commit rejected `private` and was right on this point; the second widened it to fix the membership-site case. Both goals are reachable — gate `private` behind a read-capability check rather than admitting it unconditionally.

### Minor

- **Dedup is TOCTOU.** The `$existing` scan and `add_post_meta()` aren't atomic, so two concurrent submissions from one IP both pass and both insert. Keying each row as its own meta key (e.g. `post-rating-<hash(ip)>`) would make it atomic *and* drop the O(n) unserialise-and-scan this adds to every write.
- **Empty IP becomes a shared bucket.** `get_the_ip()` returns `''` when `REMOTE_ADDR` is absent or invalid; `bsf_add_rating()` then stores `user_ip => ''` and dedups all such visitors against each other. Rejecting the submission when IP resolution fails is cleaner than storing an empty key.
- **The regenerated `.pot` ships local paths from the generating machine** — 569 lines reference a hidden worktree directory that doesn't exist for anyone else, with stale line numbers, in a translator-facing file. Pre-existing from 1.7.8, but since this PR regenerates the POT it's the natural place to drop them.
- **Version headers not bumped.** Both changelogs add `1.7.9`, but `index.php` still declares `Version: 1.7.8` and `readme.txt` `Stable tag: 1.7.8`. Fine if you bump at release time — just confirming it's deliberate.

### Verified correct

- Hidden `ip` field removal is clean: no reader of `$_POST['ip']` remains, and the JS submits via `serialize()`, so dropping the field is inert. This also closes the unescaped-output path in `bsf_display_rating()`.
- The `6`/`7`/`8` allowlist is right and complete — those are exactly the three branches that call `bsf_do_rating()` (`functions.php:578`, `:679`, `:771`).
- Rightmost-XFF reasoning is sound for the local-proxy case, and spoofed headers are correctly ignored on direct-to-origin (1b/2b above).
- Restoring the `update_post_meta()` return check, the 1–5 range bound, and the object-validation gate all do what the description claims.

**Review by sushmak02 — CHANGES_REQUESTED** [review]
Requesting changes to supersede my earlier approval at `1ac332f` — full detail in the review notes comment above.

**Blocking:** the `get_the_ip()` rewrite keys every visitor to the CDN edge IP on any CDN-fronted site (`REMOTE_ADDR` from a CDN edge is public, so the private/reserved heuristic never fires; with a private load balancer in the chain, rightmost-XFF resolves to the edge instead). Combined with the new one-rating-per-IP rule the rating feature degrades to a single globally-mutable rating — later visitors see a stranger's stars pre-selected and overwrite that row, and `aggregateRating` reports 1 review permanently. This is a regression against 1.7.8 behaviour on a very common deployment, and the changelog claim that CDNs are detected automatically isn't accurate.

**Also wants addressing before merge:** `bsf_is_rateable_post()` admits `private` posts with no read-capability check, and the nonce action isn't post-scoped, so an unauthenticated caller can still write rating meta to private content.

The rest — object validation, range bound, server-side IP sourcing, the hidden field removal, restoring the `update_post_meta()` return check — is good and I verified it does what the description says. Happy to re-review once the IP resolution and the private-post gate are sorted.

**Comment by ShubhamGupta05** [comment]
Thanks for the thorough review, @sushmak02 — both blocking and medium points were spot on. Pushed `90b59c2` addressing them.

### Blocking — `get_the_ip()` CDN collapse → reverted
Reverted the proxy-detection rewrite and restored the original `X-Forwarded-For` resolution. The private/reserved heuristic can't detect a public CDN edge, so it was keying every visitor to the edge IP — confirmed your table exactly (visitors A and B both resolved to `172.68.44.9`). With the original resolution, distinct visitors resolve correctly again.

This is the deliberate trade-off: forwarded headers are trusted (spoofable, but functional on CDN/proxy topologies — the 1.7.8 behaviour), rather than inferring proxy-ness from private ranges. A proper trusted-proxy/header-with-hop-count approach is the right long-term direction, but it's out of scope for this security PR and shouldn't block it.

### Medium — private posts writable → `publish`-only
`bsf_is_rateable_post()` now accepts `publish` only. An unauthenticated caller can't reach a read-capability check meaningfully, and no visitor is ever shown a rating form on non-published content, so admitting `private` only opened the IDOR you flagged. Membership/private-post rating can be revisited later behind an explicit capability gate if there's demand.

### Changelog
Removed the inaccurate "CDNs are detected automatically" line.

### Retained (your verified-correct items)
Object-validation gate, 1–5 range bound, server-side IP sourcing, one-rating-per-IP dedup, hidden-field removal, and the `update_post_meta()` return check are all unchanged.

### Verification
Re-tested end-to-end on a local install (unauthenticated, against `admin-ajax.php`) after the revert:

- Distinct `X-Forwarded-For` values resolve to distinct raters (CDN case fixed).
- Duplicate from the same IP → rejected; one row retained.
- `post_id` targeting a non-existent / non-rateable / **private** post → `Invalid rating request.`, no DB write.
- Out-of-range star value → `Invalid rating value.`
- Update path updates the caller's own row in place.

### Noted, not in this round
The **minor** items — TOCTOU on the dedup scan, empty-IP bucketing, and the `.pot` shipping local worktree paths — are acknowledged and can be follow-ups; happy to fold any of them in here if you'd prefer. Version headers remain 1.7.8 intentionally (bumped at release).

Re-requesting review — thanks again.

**Review by sushmak02 — APPROVED** [review]
